src/dotty/tools/dotc/core/pickling/PickleFormat.scala

```diff
@@ -0,0 +1,454 @@
+                  IMPORT        Length qual_Term Selector*
+
+  TypeParam     = TYPEPARAM     Length NameRef Type Modifier*
+  Params        = PARAMS        Length Param*
+  Param         = PARAM         Length NameRef Type Modifier*
+  Selector      = IMPORTED      Length name_NameRef
+                  RENAMED       Length from_NameRef to_NameRef
+
+  Term          = Path
+                  IDENT                Type NameRef
+                  SELECT               qual_Term possiblySigned_NameRef
+                  SUPER         Length this_Term mixinTrait_Type?
+                  APPLY         Length fn_Term arg_Term*
+                  TYPEAPPLY     Length fn_Term arg_Type*
+                  NEW           Length cls_Type
+                  PAIR          Length left_Term right_Term
+                  TYPED         Length expr_Term ascription_Type
+                  NAMEDARG      Length paramName_NameRef arg_Term
+                  ASSIGN        Length lhs_Term rhs_Term
+                  BLOCK         Length expr_Term Stat*
+                  IF            Length cond_Term then_Term else_Term
+                  CLOSURE       Length meth_Term target_Type env_Term*
+                  MATCH         Length sel_Term CaseDef*
+                  RETURN        Length meth_ASTRef expr_Term?
+                  TRY           Length expr_Term CaseDef* finalizer_Term?
+                  THROW         Length expr_Term
+                  SEQLITERAL    Length elem_Term*
+                  JSEQLITERAL   Length elem_Term*
+                  BIND          Length boundName_NameRef patType_Type pat_Term
+                  ALTERNATIVE   Length alt_Term*
+                  UNAPPLY       Length fun_Term ImplicitArg* pat_Term*
+                  ANNOTATED     Length annot_Term underlying_Term
+                  EMPTYTREE
+
+  CaseDef       = CASEDEF       Length pat_Tree guard_Tree rhs_Tree
+  ImplicitArg   = IMPLICITARG   Length arg_Tree
+  Template      = TEMPLATE      Length parent_Tree* SelfDef? Stat*
+// if there is a primary constructor, it is the first statement in Stat*..
+  SelfDef       = Param
+  ASTRef        = Nat                          		// byte position in AST payload
+
+  Path          = Constant
+                  TERMREFdirect        sym_ASTRef
+                  TERMREFstatic        fullyQualified_NameRef
+                  TERMREFsymbol        qual_Type sym_ASTRef
+                  TERMREF              qual_Type possiblySigned_NameRef
+                  THIS          Length clsRef_Type
+                  SHARED               path_ASTRef
+
+  Constant      = UNITconst
+                  FALSEconst
+                  TRUEconst
+                  BYTEconst            Nat
+                  BYTEneg              NegNat
+                  SHORTconst           Nat
+                  SHORTneg             NegNat
+                  CHARconst            Nat
+                  INTconst             Nat
```

Comment:
In our proposal of TASTY, were we have a separate constant table, we have a huge win for sharing constant `0`. If not using a shared table for constants than at least special casing `0` could be a good idea.